### PR TITLE
WIP: Performance fixies

### PR DESCRIPTION
**Pull request by osjc**
Quite a bunch of performance related warnings. Will try to fix them.

### CODE REVIEW

**Comment by osjc** [comment]
Need to have this as an active pull request because my setup does not support gcc 4.9.x

**Comment by nerzhul** [comment]
don't do a such big PR please . Stop your devs, we are in a freeze, i prefer to have more atomic commits/PR to review.

**Comment by nerzhul** [comment]
note: some v3f are copied and the clang-tidy check doesn't understand we need to copy them, be careful.

**Comment by sfan5** [comment]
I don't think it makes sense to reference `v3s16` (6 bytes) by pointer

**Comment by paramat** [comment]
Please do this in smaller commits, large PRs are overwhelming to review and tend to require a lot of rebasing. Small commits will get reviewed and merged much faster and mostly avoid rebasing.

**Comment by osjc** [comment]
> Please split this into much smaller PRs.

Thank you for this feedback. I used to work on ptojects (in my job) which insisted on "as few commits as possible" and was not exactly sure whether this is the case here as well. Glad to see the people here have much more brain than these corporate entities.

**Comment by osjc** [comment]
And, by the way this rule also makes the code history "git bisect friendly". That is a pretty useful gadget to find nasty bugs like #8090 so kudos to the leaders of this project.

**Comment by osjc** [comment]
So, here you go, this is the first commit in this megaqueue: #8096
Will keep extracting stuff that I consider completed out of this and post more numbers here.
Also it is great as each commit will get checked by travis-ci. I did not like the fact that this thing is checked as one huge lump. Please, wait with your merging until travis-ci is ok (but if your review produces comments, you can of course post them straight away ;) ).

File: src/craftdef.cpp [hunk 1]
```diff
@@ -123,6 +124,7 @@ static std::vector<std::string> craftGetItemNames(
 		const std::vector<ItemStack> &items, IGameDef *gamedef)
 {
 	std::vector<std::string> result;
+	result.reserve(64);
```
[the remark this answers is not in the document]
Comment (osjc) [reply src/craftdef.cpp]: Done: #8097. It turned out all three reserve calls in craftdef.cpp could be "unhardcoded" like this.

**Comment by osjc** [comment]
> note: some v3f are copied and the clang-tidy check doesn't understand we need to copy them, be careful.

I only "referentize" those instances explicitly pointed out by the clang-tidy check. I noticed that if the clang-tidy sees parts of the v3f structure referenced individually (v3f.X for example), it won't emit any warnings. Maybe code like "p.X = p.X;" will make it shut up.

But I found some pretty heavy things like the whole packets being copied over and over. I will extract this stuff and package it for easy review and merging. as @sfan5 points out, there is little sense using a pointer to refer to a 6 byte structure so I might want to use nop code to make it shut up. Or maybe do a google search about how to telll clang-tidy to ignore particular types in its checking and use that know-how to get rid of the warnings.

File: src/craftdef.cpp [hunk 2]
```diff
@@ -134,6 +136,7 @@ static std::vector<ItemStack> craftGetItems(
 		const std::vector<std::string> &items, IGameDef *gamedef)
 {
 	std::vector<ItemStack> result;
+	result.reserve(64);
```
[the remark this answers is not in the document]
Comment (osjc) [reply src/craftdef.cpp]: See #8097

File: src/content/subgames.cpp [hunk 3]
```diff
@@ -195,6 +199,7 @@ std::vector<SubgameSpec> getAvailableGames()
 {
 	std::vector<SubgameSpec> specs;
 	std::set<std::string> gameids = getAvailableGameIds();
+	specs.reserve(16);
```
[the remark this answers is not in the document]
Comment (osjc) [reply src/content/subgames.cpp]: See #8097

**Comment by osjc** [comment]
@nerzhul, @paramat: I put your request for smaller pull requests into the contributing resource: #8099 I know way too many developers with brains damaged by corporate entities which demand huge changes in large batches. I used to work in two of these and all their projects were like this. Hope this change will help some little if some of them decide to go hacking on this.

**Comment by osjc** [comment]
@nerzhul:

> Stop your devs,

I am alone :rofl:

**Comment by nerzhul** [comment]
i know but we are in freeze. Your developments can only be merged in the 5.1.0 merge window, please wait for february for more PR as we alread have a huge milestone to review :)

**Comment by paramat** [comment]
You're welcome to make PRs but since we're in freeze until 5.0.0 release we can't merge maintenance PRs until then and core devs may be too busy to review them. Just something to consider in case it helps you to delay this work.
However, your effort is much appreciated and we look forward to progressing with your PRs after 5.0.0 release.

**Comment by paramat** [comment]
See related #6808 "Cppcheck findings" your work might address some of those.

**Comment by sfan5** [comment]
IMO you can put the `Optimize vector construction code` into one separate PR and then split that into 3 (or 4?) commits, no need to create PRs for 1-line changes.

**Comment by paramat** [comment]
Stuff that is somewhat related can be combined into a single PR and, if suitable, split into a number of commits. Combining related stuff into a PR is useful to avoid too many little PRs.
If you follow MT dev for a while you'll get a feel for the way we balance and size PRs and commits.

**Comment by osjc** [comment]
I got more information about these comments:

> I don't think it makes sense to reference `v3s16` (6 bytes) by pointer

@sfan5: That would be true in C. But this is C++. And v3s16 is a class. That means the 6 bytes could get bloated all the way up to 24 bytes if your platform happens to be something like x64 and the C++ compiler decides that padding every field in every class to a multiple of 8 bytes is good for performance. And, second, if you add that these 6 bytes come with a large bunch of code like a copy constructor and then realize that making a copy to make a parameter pass by value involves an additional function call and you get to the point where referencing a 6 byte structure through a pointer is faster than making a copy. Read [C++ FQA](https://yosefk.com/c++fqa/) (and also [the rebuttal](https://gist.github.com/klmr/5423873) and [What's wrong with C++ FQA](http://www.cplusplus.com/forum/lounge/28407/) ) it may help you understand something about these nasty low level details under the hood. In my opinion of a seasoned C/C++ developer (30+ years of experience, C++ mostly bad) the FQA is true more often than not.

BTW, sfan5, this is not meant to be a derogatory comment in any way, I did read the forum for quite a time and I noticed you did some jawdropping contributions like "add support for waving leaves". The comment is meant to show you that "you will never know C++ fully - it keeps one surprising forever".

> note: some v3f are copied and the clang-tidy check doesn't understand we need to copy them, be careful.

@nerzhul: Could you supply an example? I am quite new to this codebase and I started this "tidyfication" to get a little bit "hands on" with it.

Namely, I noticed that the clang-tidy check only reports the parameter if it is only used to pass it around (delegating work with it to another procedure or something) and never "dereferenced" (code that looks at its pieces for example). If I have to "be careful" about converting such instances from a value passed parameter to a reference passed one, then the code in question is badly designed and needs to be fixed. However one thing needs to be clear: If you havea v3f parameter e.g. pos and you do e.g. m_pos = pos (to store that pos in a field of a class) and m_pos is declared to be v3f too, then it does not matter a tiny bit whether that "pos" is a reference or a value passed parameter, a copy is going to be made in either case. References are not pointers. So if all these instances where "we need to copy the v3f's" are like this, then there is no need to be careful.

The only way where I would have to be careful is when the v3f in question was updated by multiple threads. But then that piece of code needs to be converted to a sequential operation executed by a single thread because on modern system trying to work concurrently on a piece of data this small would actually kill performance.

But it turned out to be necessary to be careful and "referentize" only those v3f's that are reported by clang-tidy. I found a couple of instances where there were 2 v3f's next to each other and only one was reported by clang-tidy and I carelessly converted both but then I looked at the code in question closer and realized that the unreported v3f was used in a way that could actually kill performance so I redid that.

ToAll: Right now I gave this a pause. I got quite a bunch of ideas about how to fix the dreaded "formspec security check breaking mods" and I am taking a look into it.

**Comment by osjc** [comment]
> i know but we are in freeze. Your developments can only be merged in the 5.1.0 merge window, please wait for february for more PR as we alread have a huge milestone to review :)

No problem, this actually can wait for a while. The most critical thing I did was the "random form reject bug" which broke advtrains and possibly other mods and which got [merged](https://github.com/minetest/minetest/pull/8091). This one got the WIP label (actually two) and that is good enough for now.

The reason why I created the pull request early was to use the CI infrastructure which I cannot construct locally (1. gcc 4.9.x stubbornly refuses to compile on my Ubuntu and right now I want to work on my minetest game, not spend time getting a new compiler to cooperate, 2. my setup has only 8G RAM and most of it is used by minetest itself plus github so adding a dozen of virtual machines would slow things to a crawl).

Additionally, I am extracting bits and pieces of this into other pull requests so it may take a while before this thing shrinks to a manageable and reviewable size.

**Comment by osjc** [comment]
As you can see here, there is a lot of commits market WIP. These contain changes bunched together that either need to be splitted into smaller commits or significant stuff extracted out.

**Comment by osjc** [comment]
Closing for now (to keep the core developers less distracted) as I am busy working on another project and also got problems with a dying hard drive. Once I reach point where I have more time to finish this, I will reopen it.

**Pull request closed without merging**